common/src/main/java/com/wynntils/features/combat/QuickCastFeature.java

```diff
@@ -124,47 +114,61 @@ private void tryCastSpell(SpellUnit a, SpellUnit b, SpellUnit c) {
-        List<SpellDirection> spell = Stream.of(a, b, c)
-                .map(x -> (x == SpellUnit.PRIMARY) != isSpellInverted ? SpellDirection.RIGHT : SpellDirection.LEFT)
-                .toList();
-        for (int i = 0; i < spellInProgress.length; ++i) {
-            if (spellInProgress[i] != spell.get(i)) {
-                sendCancelReason(Component.translatable("feature.wynntils.quickCast.incompatibleInProgress"));
-                return;
-            }
+        Spell spell = new Spell(a, b, Models.Character.getClassType() == ClassType.ARCHER);
+        if (spells.contains(spell)) {
+            return;
         }
 
+        spells.offer(spell);
         lastSelectedSlot = McUtils.inventory().selected;
-        List<SpellDirection> remainder = spell.subList(spellInProgress.length, spell.size());
-        remainder.stream().map(SpellDirection::getSendPacketRunnable).forEach(SPELL_PACKET_QUEUE::add);
     }
 
     @SubscribeEvent
     public void onTick(TickEvent e) {
-        if (!Models.WorldState.onWorld()) return;
-
-        // Clear spell after the 40 tick timeout period
-        if (spellCountdown > 0 && --spellCountdown <= 0) spellInProgress = SpellDirection.NO_SPELL;
-
-        if (SPELL_PACKET_QUEUE.isEmpty()) return;
         if (--packetCountdown > 0) return;
+        if (currentSpell.isEmpty() && !pollSpell()) return;
 
         int currSelectedSlot = McUtils.inventory().selected;
         boolean slotChanged = currSelectedSlot != lastSelectedSlot;
-
         if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(lastSelectedSlot));
-        SPELL_PACKET_QUEUE.poll().run();
+        if (safeCasting.get()) {
+            currentSpell.poll().getSendPacketRunnable().run();
+        } else {
+            do {
+                currentSpell.peek().getSendPacketRunnable().run();
+            } while (currentSpell.poll() == SpellDirection.LEFT && !SPELL_PACKET_QUEUE.isEmpty());
+            while (currentSpell.peek() == SpellDirection.LEFT) {
+                currentSpell.poll().getSendPacketRunnable().run();
+            }
+        }
         if (slotChanged) McUtils.sendPacket(new ServerboundSetCarriedItemPacket(currSelectedSlot));
 
         // Waiting a few ticks is useful for avoiding lag related input-overlaps
-        if (!SPELL_PACKET_QUEUE.isEmpty()) packetCountdown = 3;
+        packetCountdown = Math.max(1, rightClickCooldown.get());
+    }
+
+    private boolean pollSpell() {
+        spells.remove(spell);
+        spell = null;
+        if (--spellCountdown > 0) return false;
+        if (spells.isEmpty()) return false;
+        SpellDirection[] current = Models.Spell.getLastSpell().length != 3
+                        && Duration.between(Models.Spell.getLastSpellUpdate(), Instant.now())
+                                        .toSeconds()
+                                < 3
+                ? Models.Spell.getLastSpell()
+                : SpellDirection.NO_SPELL;
```

Comment:
I can't understand what this is doing, and I think it is wrong.

Here is my understanding: `if (incorrectSpell) AND (recastTimeout) then getLastSpell else NO_SPELL`

This seems like that it was meant to be an `OR`? Also, perhaps extracting the two boolean statements like I did would clean up this code.

Reply:
`if (isInProgress) AND (!hasWnnycraftDiscardedTheProgress) then getProgress else noProgress`